Source/Diagnostics/ParticleIO.cpp

```diff
@@ -293,3 +294,114 @@ storePhiOnParticles ( PinnedMemoryParticleContainer& tmp,
         }
     }
 }
+
+void
+storeEMFieldsOnParticles (PinnedMemoryParticleContainer& tmp,
+    ElectromagneticSolverAlgo electromagnetic_solver_id, bool is_full_diagnostic  ) {
+
+    using PinnedParIter = typename PinnedMemoryParticleContainer::ParIterType;
+    using Dir = ablastr::fields::Direction;
+
+    WARPX_ALWAYS_ASSERT_WITH_MESSAGE(
+        electromagnetic_solver_id != ElectromagneticSolverAlgo::None ,
+        "output of the electromagnetic fields on the particles was requested, "
+        "but this is only available with an electromagnetic solver.");
+    WARPX_ALWAYS_ASSERT_WITH_MESSAGE(
+        is_full_diagnostic,
+        "Output of the electromagnetic fields on the particles was requested, "
+        "but this is only available with `diag_type = Full`.");
+
+
+    tmp.NewRealComp("Ex");
+    tmp.NewRealComp("Ey");
+    tmp.NewRealComp("Ez");
+    tmp.NewRealComp("Bx");
+    tmp.NewRealComp("By");
+    tmp.NewRealComp("Bz");
+
+    int const Ex_index = tmp.getParticleComps().at("Ex");
```

Comment:
There will be a minimal change we need to add to this PR once we merged #5481 (simplification of named SoA bookkeeping), but it will be easy to update once your PR is ready/out-of-WIP.